src/Utilities/ErrorHandling/CaptureForError.hpp

```diff
@@ -0,0 +1,102 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <ostream>
+#include <vector>
+
+namespace CaptureForError_detail {
+class CaptureForErrorBase {
+ protected:
+  CaptureForErrorBase() = default;
+  ~CaptureForErrorBase() = default;
+
+ public:
+  CaptureForErrorBase(CaptureForErrorBase&&) = delete;
+  CaptureForErrorBase(const CaptureForErrorBase&) = delete;
+  CaptureForErrorBase& operator=(CaptureForErrorBase&&) = delete;
+  CaptureForErrorBase& operator=(const CaptureForErrorBase&) = delete;
+
+  virtual void print(std::ostream& stream) const = 0;
+};
+
+class CaptureList {
+ public:
+  static CaptureList& the_list();
+
+  void register_self(const CaptureForErrorBase* capture);
+  void deregister_self(const CaptureForErrorBase* capture);
+
+  void print(std::ostream& stream);
+
+  CaptureList(CaptureList&&) = delete;
+  CaptureList(const CaptureList&) = delete;
+  CaptureList& operator=(CaptureList&&) = delete;
+  CaptureList& operator=(const CaptureList&) = delete;
+
+ private:
+  CaptureList();
+  ~CaptureList() noexcept(false);
+
+  std::vector<const CaptureForErrorBase*> captures_{};
+  bool currently_printing_ = false;
+};
+
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wnon-virtual-dtor"
+template <typename T>
+// NOLINTNEXTLINE(cppcoreguidelines-virtual-class-destructor)
+class CaptureForError : public CaptureForErrorBase {
+ public:
+  CaptureForError() = delete;
+  CaptureForError(CaptureForError&&) = delete;
+  CaptureForError(const CaptureForError&) = delete;
+  CaptureForError& operator=(CaptureForError&&) = delete;
+  CaptureForError& operator=(const CaptureForError&) = delete;
+
+  CaptureForError(const char* const name, const T& capture)
+      : name_(name), capture_(capture) {
+    CaptureList::the_list().register_self(this);
+  }
+
+  // Capturing an rvalue would lead to a dangling reference.
+  CaptureForError(const char* /*name*/, const T&& /*capture*/) = delete;
+
+  ~CaptureForError() noexcept(false) {
+    CaptureList::the_list().deregister_self(this);
+  }
+
+  void print(std::ostream& stream) const override {
+    stream << name_ << " = " << capture_ << "\n";
```

Comment:
maybe add `using ::operator<<;` to deal with STL as best as we can?

Reply:
And, maybe more important, call `disable_floating_point_exceptions()` (and re-enable)? Or maybe that's handle by `ERROR` itself

Reply:
`using ::operator<<;` won't compile unless I include something that defines such a function.  I can add it with an include of `StdHelpers.hpp` if you think that's a good idea.

`ERROR` and `ASSERT` have FPEs disabled when they call this.

Reply:
Ah okay, nah let's just leave as-is... Thanks!